README
HARMONI Simulator Pipeline

Written by Simon Zieleniewski & Sarah Kendrew

Last Edited: 28-04-16


----------
PURPOSE:

The HARMONI simulation pipeline code is designed to simulate observations using the HAMONI integral field spectrograph on the E-ELT. The pipeline takes a high spectral and spatial resolution input FITS format datacube, and processes it according to the chosen observing parameters. The code returns several output cubes:
 - a mock observed cube (with variance as first extension)
 - a background cube (with variance as first extension) (both noisy and noiseless versions returned)
 - a SNR cube

With further options to return:
 - a sky subtracted ("Reduced") cube (with variance as first extension)
 - a transmission cube
 - an object cube giving object flux only (both noisy and noiseless versions returned)

This pipeline provides a self contained and consistent way of performing detailed simulations of the performance of HARMONI. The mock observed output cubes can be analysed as if they are real observational data.


----------
FILES:

The pipeline currently uses 11 main python files:
hsim.py
TheSimulator.py
TheSimulator_GUI.py
TheSimulator_Specres.py
TheSimulator_PSFs.py
TheSimulator_Background.py
TheSimulator_Transmission.py
TheSimulator_ADR.py
TheSimulator_Loop.py
TheSimulator_Lowres.py
TheSimulator_Config.py

as well as several functions found in the src/modules/ sub directory.

There are also several data files:
radiance_resolution_0.15_angstroms_MICRONS.txt - From ESO Skycalc
transmission_resolution_0.15_angstroms_MICRONS.txt - From ESO Skycalc
EELT_mirror_reflectivity_mgf2agal.dat.txt - From ESO E-ELT DRM
HARMONIthruput.txt
DetectorQE.txt
HARMONI_R500_low_res_mode.txt


----------
DEPENDENCIES:

The following Python modules are essential to run the pipeline:
numpy - http://www.scipy.org/scipylib/download.html
scipy - http://www.scipy.org/scipylib/download.html
astropy - http://www.astropy.org/

The following are strongly recommended for optimum running:
wxPython - This is used to load the GUI. It can be downloaded from: http://www.wxpython.org/download.php
pprocess - This is used to allow parallel processing to improve performance. Download from https://pypi.python.org/pypi/pprocess


Make sure the modules can by found by your Python implementation. E.g. in a Python environment try:
>>> import numpy
>>> import astropy
>>> import scipy
>>> import wx
>>> import pprocess

there should be no errors!


----------
DIRECTORY SETUP:

The root directory contains several sub directories containing essential data files:

./src - main python functions stored here
     /modules - additional python functions stored here
./Output_cubes - output cubes are stored here
./Sim_data - Simulation data is stored here
	  /Background
		     /ASM_background - radiance_resolution_0.15_angstroms_MICRONS.txt stored here
	  /Throughput - EELT_mirror_reflectivity_mgf2agal.dat.txt, HARMONIthruput.txt, DetectorQE.txt stored here
		     /ASM_throughput - transmission_resolution_0.15_angstroms_MICRONS.txt stored here
	  /PSFs
	       /SCAO - SCAOdata.txt stored here
	       /LTAO - LTAOdata.txt stored here
          /R500 - HARMONI_R500_mode_data.txt stored here


----------
HOW TO USE:

The code can be run in two ways:

1. Running the GUI:
$ python hsim.py
This opens the graphical interface from which you can enter the various parameters and start the simulation.

2. From the command line:
$ python hsim.py -c arg1 arg2 arg3 etc…

Use:
$ python hsim.py -c
to display arguments list
	
The commands must be entered in the following order:
        1. datacube: Input datacube filepath
        2. DIT: Exposure time [s]
        3. NDIT: No. of exposures
        4. band: Photometric band [V+R, Iz+J, H+K, V, R, Iz, J, H, K, V-high, R-high, z, J-high, H-high, K-high, None]
        5. x-spax: x spatial pixel (spaxel) scale [mas]
        6. y-spax: y spatial pixel (spaxel) scale [mas]
	7. telescope: Telescope type (E-ELT, VLT)
	8. AO: AO mode (LTAO, SCAO, Gaussian)
        9. seeing: Atmospheric seeing FWHM [arcsec] - Between 0.67"-1.10"
        10. zenith_ang: Zenith angle [deg]
        11. user_PSF: Path to user uploaded PSF file - Enter None if not required
        12. PSF blur: Additional telescope jitter [mas]
        13. Site_temp: Site/telescope temperature [K]
        14. Spec_Nyquist: (True/False) - Set spectral sampling to Nyquist sample spectral resolution element
        15. Spec_samp: Spectral sampling [A/pix] - Only used if Spec_Nyquist = False, but enter a value regardless!
        16. Noise_force_seed: Force random number seed to take a set value (0=No, 1-2=yes and takes that value)
        17. Remove_background: (True/False) - Subtract background spectrum
	18. Return_object: (True/False) - Return object cube
	19. Return_transmission: (True/False) - Return transmission cube
	20. Turn ADR off: (True/False)

Use -h or --help to display help message and exit
Use -c or --cline option to use command line.
Use -o or --odir when using command line to specify output file directory. Default is /hsim/Output_cubes/
Use -p or --proc when using command line to specify number of cores for parallel processing. Default is N_PROC-1 (if N_PROC>1)

An example command line entry:
$ python hsim.py -c -p 5 -o /Path/to/output_dir/ /Path/to/Input_cubes/Input_cube.fits 900 20 H+K 20. 20. E-ELT LTAO 0.67 0. None 0. 280.5 True 1. 0 True True True


----------
INPUT CUBES:

The pipeline takes an input FITS file datacube containing spatial and spectral information about the astrophysical source of interest. The format should be an (x,y,lambda) datacube with each pixel containing the value of flux per sq. arcsec (e.g. J/s/m^2/A/arcsec^2).

The following FITS headers are essential:

NAXIS1, NAXIS2, NAXIS3 - Size of the x, y, and wavelength axes respectively.
CTYPE1, CTYPE2, CTYPE3 - Axis type. Accepted values: [x, y, WAVELENGTH] or [RA, DEC, WAVELENGTH].
CUNIT1, CUNIT2, CUNIT3 - Units for each axis. Accepted values: [mas, mas, microns], [mas, mas, angstroms], [mas, mas, metres], [mas, mas, nm].
CDELT1, CDELT2, CDELT3 - Sampling for each axis in units given by CUNIT1/2/3 headers.
CRVAL3 - Value of array element given by CRPIX3 in units of CUNIT3.
CRPIX3 - Denotes the array element for CRVAL3. Should be set to 1 so CRVAL3 denotes first wavelength value.
FUNITS (or BUNIT) - Flux units for each element in datacube. Accepted values: [J/s/m2/um/arcsec2], [J/s/m2/A/arcsec2], [erg/s/cm2/um/arcsec2], [erg/s/cm2/A/arcsec2].
SPECRES - Spectral resolution FWHM in units of CUNIT3.

All other headers will be ignored by the pipeline so extra information can be added.

N.B, please make sure all header keys do not have white space in them! The FITS handling modules used break down when they find header keys with spaces in. Please use underscores instead.


----------
OPTIONAL INPUT PSF IMAGE/CUBE:

The pipeline can take an input PSF FITS file instead of using the built-in PSFs. The format should either be a 2D image of a single PSF, or a 3D cube containing PSFs covering the full wavelength range covered by the input datacube.

The following FITS header are essential:

2D PSF FITS image:

NAXIS1, NAXIS2 - Size of the x, y axes respectively.
CTYPE1, CTYPE2 - Axis type. Accepted values: [x, y] or [RA, DEC].
CUNIT1, CUNIT2 - Units for each axis. Accepted values: [mas, mas]], [arcsec, arcsec].
CDELT1, CDELT2 - Sampling for each axis in units given by CUNIT1/2 headers.

3D PSF FITS cube:

NAXIS1, NAXIS2, NAXIS3 - Size of the x, y, and wavelength axes respectively.
CTYPE1, CTYPE2, CTYPE3 - Axis type. Accepted values: [x, y, WAVELENGTH] or [RA, DEC, WAVELENGTH].
CUNIT1, CUNIT2, CUNIT3 - Units for each axis. Accepted values: [mas, mas, microns], [mas, mas, angstroms], [mas, mas, metres], [mas, mas, nm].
CDELT1, CDELT2, CDELT3 - Sampling for each axis in units given by CUNIT1/2/3 headers.
CRVAL3 - Value of array element given by CRPIX3 in units of CUNIT3.
CRPIX3 - Denotes the array element for CRVAL3. Should be set to 1 so CRVAL3 denotes first wavelength value.







hsim/changelog
changes from v.114:

Fixed bug affecting simulations run on 'single channel'
Clarified how to run on multiple processors from command line
Clarified how to set output directory from command line
Added input filename into output fits cube 
Fixed PSF kink due to linear PSF interpolation
Fixed integer array indices for numpy > 1.13
Updated HARMONI and detector transmission files
Updated ELT mirror reflectivity 18 months
Updated IR and optical detector QE
Updated grating spectral resolutions and wavelength ranges
Added information to README about non-axisymmetric PSFs
Added explicit check if no PSF is selected 

---------------------

additional changes to v.115
Fixed version number and autgor details in PKG-INFO


hsim/hsim.py
'''Front-end code for HARMONI simulator
This handles the GUI and command line interfaces.

Author: Simon Zieleniewski

Last updated: 09-09-15

'''


import getopt
import sys
import os
import numpy as n
import multiprocessing as mp
from src.TheSimulator import main
import commands
from src.modules.misc_utils import path_setup



if __name__=="__main__":

    #Get version number
    try:
        verpath = path_setup('../../')
        verinfo = open(verpath+'PKG-INFO')
        ver = verinfo.readlines()[2][:-1].split('Version: ')[1]
        print 'VERSION = ', ver
        verinfo.close()
    except:
        ver = str(commands.getoutput("svnversion -c ./ | sed -e 's/[MS]//g' -e 's/^[[:digit:]]*://'"))
    
    
    optlist, args = getopt.getopt(sys.argv[1:], 'hcp:o:', ['help', 'cline', 'proc', 'odir'])

    for o, a in optlist:
        if o in ("-h", "--help"):
            print""
            print '    --------     '
            print 'HARMONI Simulator'
            print '    --------     '
            print 'SVN Version: ', ver
            print"---"
            print 'TO RUN'
            print 'GUI'
            print '>>> python hsim.py'
            print""
            print 'Command line'
            print '>>> python hsim.py -c arg arg2 ...'
            print"---"
            print 'OPTIONS'
            print '-h or --help = display this message and exit'
            print '-c or --cline = use command line. Use: >>> python hsim.py -c to display arguments list'
            print '-p or --proc = set the number of processors when using command line (1-'+str(mp.cpu_count())+')'
            print '-o or --odir = set the output file directory when using the command line (default: /hsim-#/Output_cubes)'
            print""
            sys.exit()

    nprocs = mp.cpu_count()-1           
    for o, a in optlist:
        if o in ("-p", "--proc"):
            nprocs = int(a)
            if nprocs > mp.cpu_count():
                print 'Only '+str(mp.cpu_count())+' CPUs. Using '+str(mp.cpu_count())
                nprocs = mp.cpu_count()
                
    odir = path_setup('../../Output_cubes/')         
    for o, a in optlist:
        if o in ("-o", "--odit"):
            if os.path.exists(a) and os.path.isdir(a):
                odir = a
            elif not os.path.exists(a):
                print "OUTPUT DIRECTORY DOESN'T EXIST (OR IS NOT A DIRECTORY)! PLEASE CHOOSE AGAIN. EXITING"
                sys.exit()
            
    for o, a in optlist:
        if o in ("-c", "--cline") and len(args) != 20:
            print""
            print 'COMMAND LINE USAGE'
            print ""
            print 'Enter command line arguments in following order:'
            print '1. datacube: Input datacube filepath'
            print '2. DIT: Exposure time [s]'
            print '3. NDIT: No. of exposures'
            print '4. grating: [V+R, Iz+J, H+K, Iz, J, H, K, z, J-high, H-high, K-high, None]'
            print '5. x-spax: x spatial pixel (spaxel) scale [mas]'
            print '6. y-spax: y spatial pixel (spaxel) scale [mas]'
            print '7. Telescope: Telescope type (E-ELT, VLT)'
            print '8. AO: AO mode (LTAO, SCAO, Gaussian)'
            print '9. Seeing: Atmospheric seeing FWHM [arcsec] - Between 0.67"-1.10"'
            print '10. Zenith angle: Zenith angle [deg]'
            print '11. User uploaded PSF: Path to user uploaded PSF FITS file - Enter None if not required'
            print '12. PSF blur: Additional telescope PSF blur [mas]'
            print '13. Site temp: Site/telescope temperature [K]'
            print '14. Spec_Nyquist: (True/False) - Set spectral sampling to Nyquist sample spectral resolution element'
            print '15. Spec samp: Spectral sampling [A/pix] - Only used if Spec_Nyquist=False, but enter a value regardless!'
            print '16. Noise force seed: Force random number seed to take a set value (0=No, 1-10=yes and takes that value)'
            print '17. Remove background: (True/False) - Subtract background spectrum'
            print '18. Return object cube: (True/False) - Return object FITS cube'
            print '19. Return transmission cube: (True/False) - Return transmission FITS cube'
            print '20. Turn ADR off: (True/False) - atmospheric differential refraction'
            print ""
            sys.exit()
        elif o in ("-c", "--cline") and len(args) == 20:
            print args
            datacube = str(args[0])
            DIT = int(args[1])
            NDIT = int(args[2])
            grat = str(args[3])
            spax = (float(args[4]), float(args[5]))
            telescope = str(args[6])
            AO = str(args[7])
            seeing = float(args[8])
            zenith_ang = float(args[9])
            user_PSF = str(args[10])
            res_jitter = float(args[11])
            site_temp = float(args[12])
            combine_ndits = True
            spec_Nyquist = args[13]
            if spec_Nyquist == 'True':
                spec_N = True
            elif spec_Nyquist == 'False':
                spec_N = False
            Spec_samp = float(args[14])
            noise_force_seed = int(float(args[15]))
            remove_background = args[16]
            return_obj = args[17]
            return_tra = args[18]
            return_adr = args[19]
            #Start main function
            main(datacube, odir, DIT, NDIT, grat, spax, seeing, zenith_ang, telescope,
                 user_PSF, AO, res_jitter, site_temp, combine_ndits, spec_N, Spec_samp,
                 noise_force_seed, remove_background, return_obj, return_tra, return_adr,
                 ver, nprocs)
            
##    else:
##        print""
##        print 'HARMONI Simulator'
##        print""
##        print "Unknown option keys. Use either -h for help, or -c for command line usage"
##        print""
##        sys.exit()
        

    #Use GUI interface if no command line option
    wxfound = 0
    if len(optlist) == 0 and len(args) == 0:
        try:
            import wx
            wxfound = 1
            class Form1(wx.Frame):
                
                def __init__(self, parent, title):
                    super(Form1, self).__init__(parent, title=title, 
                        size=(1120, 390))

                    self.InitUI()
                    self.Centre()
                    self.Show()

                def InitUI(self):

                    panel = wx.Panel(self)
                    titlefont = wx.Font(18, wx.DEFAULT, wx.NORMAL, wx.BOLD)

                    # Set up the menu.
                    filemenu  = wx.Menu()
                    menuAbout = filemenu.Append(wx.ID_ABOUT, "&About"," Information about this program")
                    menuExit  = filemenu.Append(wx.ID_EXIT,"&Exit"," Terminate the program")

                    # Creating the menubar.
                    menuBar = wx.MenuBar()
                    menuBar.Append(filemenu,"&File") # Adding the "filemenu" to the MenuBar
                    self.SetMenuBar(menuBar)  # Adding the MenuBar to the Frame content.

                    # Events.
                    self.Bind(wx.EVT_MENU, self.OnExit, menuExit)
                    self.Bind(wx.EVT_MENU, self.OnAbout, menuAbout)
                    
                    hbox = wx.BoxSizer(wx.HORIZONTAL)

                    fg = wx.FlexGridSizer(8, 2, 15, 10)

                    #Instrument parameters
                    inst = wx.StaticText(panel, label='Instrument')
                    inst.SetFont(titlefont)        
                    subinst = wx.StaticText(panel, label=' ')
                    INPUTCUBE = wx.StaticText(panel, label="Input Cube")
                    self.INPUTCUBEVAL = wx.FilePickerCtrl(panel, path="")
                    DIT = wx.StaticText(panel, label="DIT [s]")
                    self.DITVAL = wx.TextCtrl(panel, value='900')
                    NDIT = wx.StaticText(panel, label="NINT")
                    self.NDITVAL = wx.TextCtrl(panel, value='1')
                    XSPAX = wx.StaticText(panel, label="X Scale [mas]")
                    self.XSPAXVAL = wx.TextCtrl(panel, value='20')
                    YSPAX = wx.StaticText(panel, label="Y Scale [mas]")
                    self.YSPAXVAL = wx.TextCtrl(panel, value='20')
                    PHOTOBAND = wx.StaticText(panel, label='Grating')
                    self.PHOTOBANDVAL = wx.Choice(panel, choices=['V+R [0.47-0.82 um] (R=3100)', 'Iz+J [0.8-1.36 um]  (R=3300)', 'H+K [1.45-2.45 um] (R=3300)',
                                                                  'Iz [0.83-1.05 um] (R=7100)','J [1.05-1.32 um] (R=7100)', 'H [1.43-1.81 um] (R=7100)', 'K [1.95-2.46 um] (R=7100)',
                                                                  'z [0.82-0.91 um] (R=17200)','J-high [1.19-1.30 um] (R=17200)','H-high [1.54-1.67 um] (R=17200)','K-high [2.09-2.28 um] (R=17200)',
                                                                  'None'])
                    DIR = wx.StaticText(panel, label='Output Dir')
                    self.DIRVAL = wx.DirPickerCtrl(panel, path=path_setup('../../Output_cubes/'))
                    
                    fg.AddMany([(inst), (subinst), (INPUTCUBE),
                                (self.INPUTCUBEVAL, 1, wx.EXPAND),
                                (DIR), (self.DIRVAL, 1, wx.EXPAND),
                                (DIT), (self.DITVAL, 1, wx.EXPAND),
                                (NDIT), (self.NDITVAL, 1, wx.EXPAND),
                                (XSPAX), (self.XSPAXVAL, 1, wx.EXPAND),
                                (YSPAX), (self.YSPAXVAL, 1, wx.EXPAND),
                                (PHOTOBAND), (self.PHOTOBANDVAL, 1, wx.EXPAND),])
                    
                    fg.AddGrowableCol(1,1)

                    hbox.Add(fg, proportion=1, flag=wx.ALL|wx.EXPAND, border=10)
                       
                    #Telescope parameters(AO mode, seeing, zenith angle etc):
                    fgs = wx.FlexGridSizer(8, 2, 15, 10)

                    tele = wx.StaticText(panel, label='Telescope')
                    tele.SetFont(titlefont)
                    subtele = wx.StaticText(panel, label=' ')
                    TELESCOPE = wx.StaticText(panel, label="Telescope:")
                    self.TELESCOPEVAL = wx.Choice(panel, choices=['E-ELT', 'VLT'])
                    AOMODE = wx.StaticText(panel, label="AO Mode:")
                    self.AOMODEVAL = wx.Choice(panel, choices=['LTAO', 'SCAO', 'Gaussian'])
                    SEEING = wx.StaticText(panel, label="Zenith Seeing [arcsec]:")
                    self.SEEINGVAL = wx.TextCtrl(panel, value='0.67')
                    ZENITHA = wx.StaticText(panel, label="Zenith Angle [deg]")
                    self.ZENITHAVAL = wx.TextCtrl(panel, value='0')
                    USERPSF = wx.StaticText(panel, label='User PSF (replaces AO choice)')
                    self.USERPSFVAL = wx.FilePickerCtrl(panel, path="None")
                    gapa = wx.StaticText(panel, label=' ------------ ')
                    gapb = wx.StaticText(panel, label=' ')
                    SITETEMP = wx.StaticText(panel, label="Telescope Temperature [K]:")
                    self.SITETEMPVAL = wx.TextCtrl(panel, value='280.5')

                    fgs.AddMany([(tele), (subtele), (TELESCOPE),
                                 (self.TELESCOPEVAL, 1, wx.EXPAND),
                                 (AOMODE), (self.AOMODEVAL, 1, wx.EXPAND),
                                 (SEEING), (self.SEEINGVAL, 1, wx.EXPAND),
                                 (ZENITHA), (self.ZENITHAVAL, 1, wx.EXPAND),
                                 (gapa), (gapb),
                                 (USERPSF), (self.USERPSFVAL, 1, wx.EXPAND),
                                 (SITETEMP), (self.SITETEMPVAL, 1, wx.EXPAND)])
                    
                    fgs.AddGrowableCol(1,1)

                    hbox.Add(fgs, proportion=1, flag=wx.ALL|wx.EXPAND, border=10)
                    
                    #Misc. parameters
                    fgss = wx.FlexGridSizer(10, 2, 15, 10)

                    misc = wx.StaticText(panel, label="Miscellaneous")
                    misc.SetFont(titlefont)
                    submisc = wx.StaticText(panel, label=' ')
                    REMOVE_BG = wx.StaticText(panel, label='Subtract Background')
                    self.REMOVE_BGVAL = wx.CheckBox(panel)
                    OUTPUT_OB = wx.StaticText(panel, label='Return Object Cube')
                    self.OUTPUT_OBVAL = wx.CheckBox(panel)
                    OUTPUT_TS = wx.StaticText(panel, label='Return Transmission Cube')
                    self.OUTPUT_TSVAL = wx.CheckBox(panel)
                    gapaa = wx.StaticText(panel, label=' ------------ ')
                    gapbb = wx.StaticText(panel, label=' ')
                    NOISESEED = wx.StaticText(panel, label="Noise Seed")
                    self.NOISESEEDVAL = wx.Choice(panel, choices=['Random', '1', '2'])
                    SETR = wx.StaticText(panel, label="Set Spec Samp [A/pix]")
                    self.SETRVAL = wx.TextCtrl(panel)
                    N_PROC = wx.StaticText(panel, label='No. of processors (1-'+str(mp.cpu_count())+')')
                    self.N_PROCVAL = wx.TextCtrl(panel, value=str(mp.cpu_count()-1))
                    RESJIT = wx.StaticText(panel, label="Additional PSF Blur [mas]:")
                    self.RESJITVAL = wx.TextCtrl(panel, value='0')
                    ADR = wx.StaticText(panel, label="Turn off ADR")
                    self.ADRVAL = wx.CheckBox(panel)

                    fgss.AddMany([(misc), (submisc),
                                  (REMOVE_BG), (self.REMOVE_BGVAL, 1, wx.EXPAND),
                                  (OUTPUT_OB), (self.OUTPUT_OBVAL, 1, wx.EXPAND),
                                  (OUTPUT_TS), (self.OUTPUT_TSVAL, 1, wx.EXPAND),
                                  (gapaa), (gapbb),
                                  (N_PROC), (self.N_PROCVAL, 1, wx.EXPAND),
                                  (NOISESEED), (self.NOISESEEDVAL, 1, wx.EXPAND),
                                  (SETR), (self.SETRVAL, 1, wx.EXPAND),
                                  (RESJIT), (self.RESJITVAL, 1, wx.EXPAND),
                                  (ADR), (self.ADRVAL, 1, wx.EXPAND)])

                    fgss.AddGrowableCol(1,1)

                    hbox.Add(fgss, proportion=1, flag=wx.ALL|wx.EXPAND, border=10)

                    panel.SetSizer(hbox)

                    # A button
                    button =wx.Button(panel, 10, "Commence Simulation", wx.Point(465, 325))
                    wx.EVT_BUTTON(panel, 10, self.OnClick)

                def OnClick(self,event):       
                    #Extract parameters for simulation run:
                    cubefile = str(self.INPUTCUBEVAL.GetPath())
                    ditval = float(self.DITVAL.GetValue())
                    nditval = float(self.NDITVAL.GetValue())
                    xspaxval = float(self.XSPAXVAL.GetValue())
                    yspaxval = float(self.YSPAXVAL.GetValue())
                    photoband = str(self.PHOTOBANDVAL.GetStringSelection()).split(' ')[0]
                    telescopeval = str(self.TELESCOPEVAL.GetStringSelection())
                    aomodeval = str(self.AOMODEVAL.GetStringSelection())
                    seeingval = float(self.SEEINGVAL.GetValue())
                    zenithaval = float(self.ZENITHAVAL.GetValue())
                    resjitval = float(self.RESJITVAL.GetValue())
                    sitetempval = float(self.SITETEMPVAL.GetValue())
                    noiseseedval = self.NOISESEEDVAL.GetStringSelection()
                    if noiseseedval == 'Random':
                        noiseseedval = 0.
                    setrchoice = self.SETRVAL.GetValue()
                    if setrchoice:
                        spec_nyq = False
                        setrval = float(setrchoice)
                    elif not setrchoice:
                        spec_nyq = True
                        setrval = 1.
                    remove_bgval = str(self.REMOVE_BGVAL.GetValue())
                    return_obval = str(self.OUTPUT_OBVAL.GetValue())
                    return_tsval = str(self.OUTPUT_TSVAL.GetValue())
                    return_adrval = str(self.ADRVAL.GetValue())
                    nprocs = int(self.N_PROCVAL.GetValue())
                    odir = str(self.DIRVAL.GetPath())
                    user_PSF = str(self.USERPSFVAL.GetPath())
                    combnditsval=True

                    print 'Filename: ', cubefile
                    print 'DIT = ', ditval
                    print 'NINT = ', nditval
                    print 'X spaxels = ', xspaxval
                    print 'Y spaxels = ', yspaxval
                    print 'Grating = ', photoband
                    print 'Telescope: ', telescopeval
                    print 'AO = ', aomodeval
                    print 'Seeing = ', seeingval
                    print 'Zenith angle = ', zenithaval
                    print 'Residual jitter = ', resjitval
                    print 'Temperature = ', sitetempval
                    print 'Noise seed = ', noiseseedval
                    print 'Spectral Nyquist sampling? ', spec_nyq
                    print 'Subtract background? ', remove_bgval
                    print 'User PSF? ' , user_PSF
                    print 'Return object cube? ', return_obval
                    print 'Return transmission? ', return_tsval
                    print 'ADR off? ', return_adrval
                    print 'No. of processors = ', nprocs
                    print 'Output directory = ', odir
                           
                    #start main program
                    main(cubefile, odir, ditval, nditval, photoband, (xspaxval, yspaxval),
                         seeingval, zenithaval, telescopeval, user_PSF, aomodeval, resjitval,
                         sitetempval, combnditsval, spec_nyq, setrval, noiseseedval, remove_bgval,
                         return_obval, return_tsval, return_adrval, ver, nprocs)

                def OnExit(self,e):
                    self.Close(True)  # Close the frame.

                def OnAbout(self,e):
                    info = wx.AboutDialogInfo()
                    #info.SetIcon(wx.Icon('hunter.png', wx.BITMAP_TYPE_PNG))
                    info.SetName('HSIM')
                    info.SetVersion(ver)
                    info.SetDescription('HARMONI simulation pipeline')
                    #info.SetCopyright('(C) 2014')
                    info.SetWebSite('http://www-astro.physics.ox.ac.uk/instr/HARMONI/Simulator/simulator.html')
                    #info.SetLicence(licence)
                    info.AddDeveloper('Simon Zieleniewski, Sarah Kendrew, Ryan Houghton & Niranjan Thatte')
                    info.AddDocWriter('Simon Zieleniewski')
                    #info.AddArtist('Artist')
                    #info.AddTranslator('Translator')
                    wx.AboutBox(info)

        except:
            print""
            print 'HARMONI Simulator'
            print '    --------     '
            print 'NO WX MODULE'
            print 'wxPython can be downloaded from: http://www.wxpython.org/'
            print""
            print 'COMMAND LINE USAGE ONLY'
            print""
            print '>>> python hsim.py -c arg1 arg2 ...'
            print""
            print 'To see help:'
            print""
            print '>>> python hsim.py -c -h'
            print""
            print 'To see required arguments:'
            print""
            print '>>> python hsim.py -c'
            print""
            print"---"
            sys.exit()
        if wxfound:
            #GUI interface
            app = wx.App()
            Form1(None, title="HARMONI Simulator Interface")
            app.MainLoop()



